src/mango.ts

```diff
@@ -0,0 +1,491 @@
+    const parsedEvent = JSON.parse(event.data);
+    const {
+      owner, clientOrderId, quantity, price,
+    } = parsedEvent.event;
+    if (owner !== mangoAccount.publicKey.toBase58()) {
+      return;
+    }
+    if (clientOrderId === gammaBidID || clientOrderId === gammaAskID) {
+      gammaFillQty += Math.abs(quantity);
+      // Once the gamma fills have crossed the threshold, reset the orders.
+      if (gammaFillQty > netGamma * GAMMA_COMPLETE_THRESHOLD_PCT) {
+        console.log(
+          scalper.symbol,
+          'Gamma Filled',
+          gammaFillQty,
+          clientOrderId === gammaBidID ? 'BOUGHT FOR' : 'SOLD AT',
+          price,
+          new Date().toUTCString(),
+        );
+        fillFeed.removeEventListener('message', gammaFillListener);
+        // Do not need to remove the unfilled order since it will be cancelled in
+        // the recursive call.
+        gammaScalpMango(
+          dipProduct,
+          scalper,
+          mangoClient,
+          mangoAccount,
+          mangoGroup,
+          perpMarket,
+          fillFeed,
+          gammaScalpCount + 1,
+        );
+      } else {
+        console.log('Gamma Partially Filled', gammaFillQty, 'of', netGamma);
+      }
+    }
+  };
+
+  if (fillFeed.readyState === WebSocket.OPEN) {
+    fillFeed.addEventListener('message', gammaFillListener);
+    console.log(scalper.symbol, 'Listening For gamma scalps');
+  } else {
+    console.log(scalper.symbol, 'Websocket State', fillFeed.readyState);
+  }
+
+  // Place gamma scalp bid & offer.
+  try {
+    const gammaBidTx = await mangoClient.perpPlaceOrderIx(
+      mangoGroup,
+      mangoAccount,
+      perpMarket.perpMarketIndex,
+      PerpOrderSide.bid,
+      gammaBid,
+      netGamma,
+      undefined,
+      gammaBidID,
+      PerpOrderType.postOnlySlide,
+    );
+    const gammaAskTx = await mangoClient.perpPlaceOrderIx(
+      mangoGroup,
+      mangoAccount,
+      perpMarket.perpMarketIndex,
+      PerpOrderSide.ask,
+      gammaAsk,
+      netGamma,
+      undefined,
+      gammaBidID,
+      PerpOrderType.postOnlySlide,
+    );
+    const gammaOrdersTx: TransactionInstruction[] = [gammaBidTx, gammaAskTx];
+    console.log(scalper.symbol, 'Gamma Bid', gammaBid, 'ID', gammaBidID);
+    console.log(scalper.symbol, 'Gamma Ask', gammaAsk, 'ID', gammaAskID);
+    await mangoClient.sendAndConfirmTransaction(gammaOrdersTx);
+  } catch (err) {
+    console.log(scalper.symbol, 'Gamma Error', err, err.stack);
+  }
+  console.log(`${scalper.symbol} Gamma Spread % ${((gammaAsk - gammaBid) / fairValue) * 100} Liquidity $ ${netGamma * 2 * fairValue}`);
+
+  // Sleep for the max time of the reruns then kill thread
+  await sleepExact(SCALPER_WINDOW_SEC);
+  console.log(scalper.symbol, 'Remove stale gamma fill listener', gammaBidID, gammaAskID);
+  fillFeed.removeEventListener('message', gammaFillListener);
+}
+
+export async function runMangoScalper(dipProduct: DIPDeposit[], scalper: Scalper) {
```

Comment:
rename to closer to what it does or move all of the logic that could result in calling openbook into a helper function that is called in the caller first

Reply:
loadMangoAndPickScalper added